tools/codesign_mac_executable.py

```diff
@@ -0,0 +1,232 @@
+#
+# Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+#
+
+# Please refer to https://medium.com/@parthnaik92/codesign-and-notarize-unreal-engine-mac-builds-for-distribution-outside-of-the-mac-store-d2f6e444f3a7
+# for pre-requisites and setup your system before trying to run this file.
+
+import argparse
+import os
+import shutil
+import subprocess
+import time
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--macos_executable", required=True)
+    parser.add_argument("--entitlements_file", required=True)
+    parser.add_argument("--developer_id", required=True)
+    parser.add_argument("--output_dir", required=True)
+    parser.add_argument("--apple_username", required=True)
+    parser.add_argument("--apple_password", required=True)
+    parser.add_argument("--wait_time_seconds", type=float, default=600.0)
+    args = parser.parse_args()
+
+    assert os.path.exists(args.macos_executable)
+    radio_effect_units = [
+        os.path.join(args.macos_executable, "Contents", "Resources", "RadioEffectUnit.component"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Build", "Mac", "RadioEffectUnit")]
+
+    print(f"[SPEAR | codesign_mac_executable.py] Removing {radio_effect_units[0]}.")
+    shutil.rmtree(radio_effect_units[0], ignore_errors=True)
+
+    print(f"[SPEAR | codesign_mac_executable.py] Removing {radio_effect_units[1]}.")
+    shutil.rmtree(radio_effect_units[1], ignore_errors=True)
+
+    print(f"[SPEAR | codesign_mac_executable.py] Changing rpaths...")
+
+    # change current working directory to add relative rpaths 
+    cwd = os.getcwd()
+    os.chdir(os.path.join(args.macos_executable, ".."))
+
+    executable_name = os.path.basename(args.macos_executable)
+
+    cmd = [
+        "install_name_tool",
+        "-add_rpath",
+        os.path.join(executable_name, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "Ogg", "Mac"),
```

Comment:
what happens when you run this part of the script repeatedly? are you adding redundant rpaths to the executable or these dylibs?

Reply:
It would through an error. Something like this

```bash
[SPEAR | sign_executable.py] Changing rpaths...
[SPEAR | sign_executable.py] Chaning working directory to /Users/rachithp/code/spear-mac-build-system/sign-app/SpearSim-Mac-Shipping.
[SPEAR | sign_executable.py] Executing: install_name_tool -add_rpath SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac/libogg.dylib
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: for: SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac/libogg.dylib (for architecture x86_64) option "-add_rpath SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac" would duplicate path, file already has LC_RPATH for: SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: for: SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac/libogg.dylib (for architecture arm64) option "-add_rpath SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac" would duplicate path, file already has LC_RPATH for: SpearSim-Mac-Shipping.app/Contents/UE4/Engine/Binaries/ThirdParty/Ogg/Mac
Traceback (most recent call last):
  File "sign_executable.py", line 68, in <module>
    assert cmd_result.returncode == 0
AssertionError
```

Reply:
In this case, copy the executable to the temp folder and work in the temp folder, so you can assume you're working with a pristine executable instead of modifying it in-place.

Reply:
Good point, changed this.